### PR TITLE
WIP Optionally install oAuth configuration

### PR DESCRIPTION
**Pull request by tilleryd**
**Please provide enough information so that others can review your pull request:**

1. Explain the **details** for making this change. What existing problem does the pull request solve? If it resolves an existing issue, be sure to use a Github [keyword](https://help.github.com/en/articles/closing-issues-using-keywords) to automatically close it.

This PR allows users to install Grey Matter with oAuth configured. Two ingresses will be setup, one for oAuth and one for PKI. In the oAuth scenario, the only service that is configured in the mesh for the standard PKI edge will be gm-control-api. This allows users to continue to hit the API via the CLI from the edge.

2. What **changes to custom.yaml** are required?

In greymatter.yaml, there are now two new edge ingress configurations using Voyager and an additional boolean to enable oAuth in the mesh.

In greymatter-secrets.yaml, there are new oAuth attributes which will be used to configure the edge-oauth proxy in the mesh.

3. Have you documented any additional setup steps or configurations options?

Will be in a follow-up PR.

4. Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI.

### CODE REVIEW

File: edge/templates/edge-ingress.yaml [hunk 1]
```diff
@@ -29,5 +29,7 @@ metadata:
 spec:
   rules:
 {{ toYaml .rules | indent 4 }}
+  tls:
```
Comment (chrisbsmith) [inline edge/templates/edge-ingress.yaml]: This assumes that TLS will always be provided.  In an nginx-ingress configuration, we enable nginx to do tls passthrough, so this the `tls` block is not needed.

Recommend adding a check here to see if the `tls` config option exists in the values file first.

File: edge/templates/edge-oauth-ingress.yaml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+kind: Ingress
+metadata:
+  name: edge-oauth
+  namespace: {{ $.Release.Namespace }}
+  annotations:
+{{ toYaml .annotations | indent 4 }}
+spec:
+  rules:
+{{ toYaml .rules | indent 4 }}
+  tls:
```
Comment (chrisbsmith) [inline edge/templates/edge-oauth-ingress.yaml]: My tls comment from above would hold true here too.

File: edge/templates/edge.yaml [hunk 3]
```diff
@@ -17,14 +17,14 @@ spec:
     spec:
       containers:
       - name: edge
-        image: {{ tpl .Values.sidecar.image $ | quote }}
-        imagePullPolicy: {{ .Values.sidecar.imagePullPolicy }}
+        image: {{ tpl .Values.edge.sidecar.image $ | quote }}
```
Comment (chrisbsmith) [inline edge/templates/edge.yaml]: This is interesting because we chose to use `sidecar.image` since edge is nothing but a sidecar configured as an edge.  However, I see that we also wrote `edge.sidecar.image`.  Part of me wants to get rid of `edge.sidecar.image` as it's just another place we need to keep track of a sidecar version.  I'd expect the edge and sidecar versions to always be in sync.  Thoughts?

File: gm-control-api/config/bootstrap.sh [hunk 4]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env sh
 
-MESH_CONFIG_DIR="/etc/config/mesh/"
+MESH_CONFIG_DIR="/etc/config/mesh"
```
Comment (chrisbsmith) [inline gm-control-api/config/bootstrap.sh]: Have you tested without the trailing `/`?

File: gm-control-api/json/special/edge-oauth/proxy.json [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  "proxy_key": "edge-oauth-proxy",
+  "domain_keys": ["edge-oauth"],
+  "listener_keys": ["edge-oauth-listener"],
+  "name": "edge-oauth",
+  "listeners": null,
+  "active_proxy_filters": ["gm.metrics", "gm.oauth"],
+  "proxy_filters": {
+    "gm_oauth": {
+      "provider": "https://deciphernow-davidtillery.okta.com",
+      "client_id": "0oa28nhtmfWgx0hTu357",
```
Comment (chrisbsmith) [inline gm-control-api/json/special/edge-oauth/proxy.json]: These probably should be secrets or env vars that can be substituted into place.

File: gm-control-api/templates/configmaps.yaml [hunk 6]
```diff
@@ -1,18 +1,18 @@
...
 {{- $svidList = append $svidList (print "spiffe://" $top.Values.global.spire.trustDomain "/spire-client/mTLS") }}
 {{- /* We loop over all the services to create the individual service objects */}}
 {{- range $name, $service := .Values.global.services }}
 {{- if $service.controlEnabled }}
----
 {{/* {{- $lSecret := (default $service.removeListenerSecret false) }} */}}
 {{- $serviceExtras := (dict "top" $top "authorizedSvids" $svidList "mTLS" $top.Values.global.spire.enabled "zone" $top.Values.gmControlApi.zone "svidList" (toJson $svidList)) }}
 {{- $svc := merge $service $serviceExtras }}
 {{- $scope := set $top "service" $svc }}
+---
```
Comment (chrisbsmith) [inline gm-control-api/templates/configmaps.yaml]: Why did you move this break?  I'm not sure of the impact it may have.

File: greymatter.yaml [hunk 7]
```diff
@@ -701,60 +704,104 @@ sidecar:
...
-      # For instance, these values will work for nginx ingress
+      #################
+      # NGINX INGRESS #
+      #################
+      apiVersion: extensions/v1beta1
+      annotations:
+        kubernetes.io/ingress.class: "nginx"
+        nginx.ingress.kubernetes.io/ssl-passthrough: "true"
+      rules:
+        - host: pki.dtillery-ingress.com
```
Comment (chrisbsmith) [inline greymatter.yaml]: don't forget that you have "personal" info in this file.

File: edge/templates/edge-ingress.yaml (hunk 1 above)
Comment (tilleryd) [reply edge/templates/edge-ingress.yaml]: removed in favor of using Voyager with SSL passthrough

File: edge/templates/edge-oauth-ingress.yaml (hunk 2 above)
Comment (tilleryd) [reply edge/templates/edge-oauth-ingress.yaml]: removed in favor of using Voyager with SSL passthrough

File: edge/templates/edge.yaml (hunk 3 above)
Comment (tilleryd) [reply edge/templates/edge.yaml]: Agreed, but that's out of scope of this PR.

File: gm-control-api/config/bootstrap.sh (hunk 4 above)
Comment (tilleryd) [reply gm-control-api/config/bootstrap.sh]: Yes, all subsequent string concatenations led to double forward slashes before this change.

File: gm-control-api/templates/configmaps.yaml (hunk 6 above)
Comment (tilleryd) [reply gm-control-api/templates/configmaps.yaml]: It's just for readability. The break is now directly above the following YAML.

Comment (tilleryd) [reply gm-control-api/templates/configmaps.yaml]: I tested everything and it works fine.

File: greymatter.yaml (hunk 7 above)
Comment (tilleryd) [reply greymatter.yaml]: OBE

**Comment by chrisbsmith** [comment]
@tilleryd This has been a WIP for quite a long time now.  At this point, I don't think that we should push it forward to the `release-2.1` branch. What do you think?

**Comment by tilleryd** [comment]
This is OBE due to the new filters and helm-charts restructure.
